torch/_dynamo/utils.py

```diff
@@ -212,12 +221,42 @@ def time_wrapper(*args, **kwargs):
             compilation_time_metrics[key].append(time_spent)
             if phase_name:
                 frame_key = str(curr_frame)
-                if frame_key not in frame_phase_timing:
-                    frame_phase_timing[frame_key] = {}
-                if phase_name not in frame_phase_timing[frame_key]:
-                    frame_phase_timing[frame_key][phase_name] = time_spent
+                # fwd only compilation stages: entire_frame_compile, backend_compile_time.
+                # use frame_key as time aggregation key.
+                if fwd_only:
+                    _add_time_spent(frame_key, phase_name, time_spent)
                 else:
-                    frame_phase_timing[frame_key][phase_name] += time_spent
+                    # fwd + bwd compilation stages: inductor_compile, code_gen.
+                    # use frame_key as time aggregation key for fwd graphs;
+                    # use compile_id as time aggregation key for bwd graphs.
+                    if torch._guards.TracingContext.try_get() is not None:
+                        aot_graph_name = str(
+                            torch._guards.TracingContext.get().aot_graph_name
+                        )
+                        if "forward" in aot_graph_name or "inference" in aot_graph_name:
+                            _add_time_spent(frame_key, phase_name, time_spent)
+                        elif "backward" in aot_graph_name:
+                            compile_id = str(
+                                torch._guards.CompileContext.current_compile_id()
+                            )
+                            _add_time_spent(compile_id, phase_name, time_spent)
+
+                            # log backward compilation metrics at the end of `inductor_compile` of bwd graph,
+                            # one record for one bwd graph.
+                            if phase_name == "inductor_compile":
+                                inductor_compile_time = frame_phase_timing[
+                                    compile_id
+                                ].get("inductor_compile", None)
+                                code_gen_time = frame_phase_timing[compile_id].get(
+                                    "code_gen", None
+                                )
+                                metrics = BwdCompilationMetrics(
+                                    compile_id,
+                                    inductor_compile_time,
+                                    code_gen_time,
+                                )
+                                record_compilation_metrics(metrics)
```

Comment:
This is copied from the code in AOTAutograd. But we shouldn't have two copies of this. Can you also delete the AOTAutograd code in this PR?

Reply:
No, this is not copied from AOT, can you advice where is the similar code in AOT? I can do refactor if there is.

Reply:
in torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py search for `aot_autograd_backward_compilation_metrics`